src/java_bytecode/ci_lazy_methods.cpp

```diff
@@ -477,17 +477,28 @@ void ci_lazy_methodst::gather_field_types(
   ci_lazy_methods_neededt &needed_lazy_methods)
 {
   const auto &underlying_type=to_struct_type(ns.follow(class_type));
-  for(const auto &field : underlying_type.components())
+  const typet &element_type=
+    static_cast<const typet &>(class_type.find(ID_C_element_type));
+  if(element_type.is_not_nil() && element_type.id()==ID_pointer)
   {
-    if(field.type().id()==ID_struct || field.type().id()==ID_symbol)
-      gather_field_types(field.type(), ns, needed_lazy_methods);
-    else if(field.type().id()==ID_pointer)
+    // This is a reference array -- mark its element type available.
+    initialize_all_needed_classes_from_pointer(
+      to_pointer_type(element_type), ns, needed_lazy_methods);
+  }
+  else
+  {
+    for(const auto &field : underlying_type.components())
     {
-      // Skip array primitive pointers, for example:
-      if(field.type().subtype().id()!=ID_symbol)
-        continue;
-      initialize_all_needed_classes_from_pointer(
-        to_pointer_type(field.type()), ns, needed_lazy_methods);
+      if(field.type().id()==ID_struct || field.type().id()==ID_symbol)
+        gather_field_types(field.type(), ns, needed_lazy_methods);
+      else if(field.type().id()==ID_pointer)
+      {
+        // Skip array primitive pointers, for example:
+        if(field.type().subtype().id()!=ID_symbol)
```

Comment:
I realise this isn't from this PR  - but is there a better way to skip array primitive pointers as this feels like a fairly loose check.